### ISSUE TITLE
Using neovim installed with bob as sudo

### ISSUE THREAD

**Issue by nfejzic** [MordechaiHadad/bob#95]
Hi! First of all, all compliments to you, this is an amazing project. 

Today I tried to use the neovim as sudo, and noticed that it does not work.

## What I tried: 

* set `SUDO_EDITOR`, `VISUAL` and `EDITOR` to `nvim`
* `sudo nvim ...` prints: 
  `sudo: nvim: command not found`
* `sudo -e ...` opens the file in default editor `nano`
* `sudoedit ...` opens the file in default editor `nano`

I'm currently using Fedora Workstation 37. 

Is there some configuration I'm missing, or is this something that generally does not work right now. If it's the latter, then please consider this as a feature request. I would also gladly try to help with the implementation of this if help is needed.

**Comment by MordechaiHadad** [comment]
@nfejzic hey that's an edge case I wasn't considering, thank you for opening this issue!

**Comment by MordechaiHadad** [comment]
Okay I don't quite get your error but this:
```
Mar 03 11:54:13.119 ERROR Error: The used file required for bob could not be found. This could mean that Neovim is not installed through bob.
```

But I get why, since the home directory changes to be the root's 🤔

**Comment by nfejzic** [comment]
> Okay I don't quite get your error but this:

Ah yes! I got that error when I tried to run the neovim using following: 

`sudo ~/.local/share/bob/nvim-bin/nvim ...`

In my case simply `sudo nvim` does not invoke that, but I assume that might depend on the way bob was installed. I installed it using cargo: `cargo install bob-nvim`.

**Comment by MordechaiHadad** [comment]
@nfejzic I managed to get sudo nvim to work, but it loads without the config, is that how normally installed neovim works?

**Comment by nfejzic** [comment]
Yes, that is how it usually works. I suppose it's also good, because we don't want to run any plug-ins with sudo.

**Comment by MordechaiHadad** [comment]
Created a PR tell me if it works for you!

**Comment by nfejzic** [comment]
> Created a PR tell me if it works for you!

Hmm, correct me if I did something wrong, but after installing with: 
`cargo install --git https://github.com/MordechaiHadad/bob.git --branch 'fix/sudo-error'`

and running bob I get following error: 

```bash
Mar 03 17:20:57.885 ERROR Error: environment variable not found
```

Some `unwrap()` causing this?

**Comment by nfejzic** [comment]
It occurred to me, that this should probably be solved a little differently than what you did in PR #96. 

Since I installed cargo using rustup, this means that all my cargo binaries are installed locally in my home directory. 

That means that `bob` can't be run with sudo privileges (`sudo: bob: command not foun`), and that means that the `$SUDO_USER` will never be set when running bob. 

A possible solution would be to ask the user if neovim should be made system available (i.e. symlink in /bin). If that's the case, sudo password is needed and then symlink is created etc.

**Comment by MordechaiHadad** [comment]
@nfejzic yeah no worries! it was just temporary code to see if SUDO_USER will work, I will use USER in case sudo user doesnt exist

**Comment by MordechaiHadad** [comment]
> > Created a PR tell me if it works for you!
> 
> Hmm, correct me if I did something wrong, but after installing with: `cargo install --git https://github.com/MordechaiHadad/bob.git --branch 'fix/sudo-error'`
> 
> and running bob I get following error:
> 
> ```shell
> Mar 03 17:20:57.885 ERROR Error: environment variable not found
> ```
> 
> Some `unwrap()` causing this?

this should only happen if running bob without sudo, I am fixing it currently!

**Comment by MordechaiHadad** [comment]
should be fixed now with new commit

**Comment by nfejzic** [comment]
Neither `sudo nvim` nor `sudo bob` work on my machine. Like I said above, the `$SUDO_USER` var will never be set in my case, because all binaries installed with `cargo` live in `~/.cargo/bin/`, and they won't get run with `sudo`. 🤔

**Comment by MordechaiHadad** [comment]
This is weird, does it only happen to you when installing from crates.io or using ` cargo install --git https://github.com/MordechaiHadad/bob.git --branch 'fix/sudo-error'` as well?
I installed bob now via cargo install using the git url and it works with no issues

**Comment by MordechaiHadad** [comment]
also does this behavior persist across all of your cargo installed binaries? on my machine I can run all of my programs installed via cargo using sudo with no issue

**Comment by nfejzic** [comment]
I tried fetching `fix/sudo-error` branch and installing with `cargo install --path ...`, also with `cargo install --git ... --branch `fix/sudo-erro` and neither did work. 

What I found out is in my `/etc/sudoers` there is `secure_path` defined, which means my `$PATH` variable is reset when running with `sudo`. 

> also does this behavior persist across all of your cargo installed binaries? on my machine I can run all of my programs installed via cargo using sudo with no issue

Yes, however there is a [workaround](https://unix.stackexchange.com/a/83194): `sudo env "PATH=$PATH" nvim`. However, in this case I get the following: 
`Mar 05 13:03:13.236 ERROR Error: The used file required for bob could not be found. This could mean that Neovim is not installed through bob.`

**Comment by MordechaiHadad** [comment]
this thing: 
```
## Uncomment to use a hard-coded PATH instead of the user's to find commands
# Defaults secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
```

also do you mind opening an issue for cargo itself and link this issue? I don't really know how cargo works exactly so it will have to grasp the situation better

**Comment by MordechaiHadad** [comment]
had a friend test installing bob from cargo works fine for him as well

**Comment by nfejzic** [comment]
> this thing?:
> 
> ```
> ## Uncomment to use a hard-coded PATH instead of the user's to find commands
> # Defaults secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
> ```
Yes, that came so by default on my OS (Fedora Workstation 37)

I did some debugging and found something out. I pulled your latest changes in `fix/sudo-erro` branch, installed bob through there and tried to print some variables out. Then I tried switching between neovim nightly and stable, however the prints did nothing. 
Finally I deleted the `nvim-bin/nvim` binary, ran `bob use stable` again, and then it worked. 

Note I ran it with `sudo env "PATH=$PATH" nvim`. I think this is expected behavior, possible workarounds are to disable `secure_path`, or to provide functionality in `bob` to create a symlink at some location available to sudo (like `/usr/bin/`), and the last one is to set `SUDO_EDITOR` to location of bob nvim binary: `SUDO_EDITOR='/home/user/.local/share/bob/nvim-bin/nvim`. I'll probably go with the last one for now. 

You could probably add this to FAQ or troubleshooting section as info. 👀

**Comment by MordechaiHadad** [comment]
Hmmm I see, why is `sudo nvim` not working even tho bob doesn't depend on `$PATH` at all?
Also thanks for the suggestion I will add a troubleshooting section, do you think it should go to the readme or is it better to create a wiki page?

**Comment by nfejzic** [comment]
> Hmmm I see, why is `sudo nvim` not working even tho bob doesn't depend on `$PATH` at all? 

The bob doesn't, but the location of `nvim` binary is not on my `$PATH` for sudo. So sudo can't find it at all.

> Also thanks for the suggestion I will add a troubleshooting section, do you think it should go to the readme or is it better to create a wiki page?

Both are fine for me... Maybe if troubleshooting is small leave it in readme, if there's a lot then in wiki.
